TagManager and Tags

fterh opened the pull request:
In progress, don't merge yet.

fterh committed:
Implement TagManager (with TagManagerInterface)

fterh committed:
Update Tag struct

* Add documentation
* Add convenience computed properties

fterh committed:
Add tests for TagManager

fterh commented:
Ready to merge.

Important: review the API (TagManagerInterface.swift) and see if it's okay :) Implementation and tests follow the interface specifications.

fterh committed:
Add .tags property 

TagManager.tags returns a dictionary mapping all parent Tags to sets of their children Tags.

fterh committed:
Fix swiftlint style violation

fterh committed:
Organize project directory /commons

fterh committed:
Add FileStorageManager to /commons/utils

fterh committed:
Update FileStorageManager methods to throw

fterh committed:
Add persistent storage to TagManager

fterh committed:
Update TagManager tests

fterh commented:
Ready to merge.

Main features:
* TagManager is implemented as semi-singleton with private init()
* To instantiate, use static method `.create(withPersistence: Bool)`
* Non-persistent TagManager for unit testing (of logic)
* Persistent TagManager (singleton) for production
* TagManager automatically saves its state to storage (persistence)
* When creating a persistent TagManager, try reconstructing from saved data (if possible); otherwise, create an empty/fresh instance of a persistent TagManager
* Logging for I/O methods (since we can't test it)

kumuwu commented on bacon/model/Tags/TagManager.swift:
```diff
@@ -0,0 +1,253 @@
+        guard !allTags.contains(childTag) else {
+            throw DuplicateTagError(message: "\(childTag) already exists")
+        }
+
```
Shall assert the parent tag already exists?

kumuwu replied:
I see that in the later part there is a method to add the parent as well when the parent is not already there. From the user side, he should never be able to add a child tag to a non-existent parent tag tho? if it happens it's probably some error within the programme?

fterh replied:
Okay, I'll change the specifications then. 

Edit: fixed in commit 81bba2

kumuwu commented on bacon/model/Tags/TagManager.swift:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    var tags: [Tag: Set<Tag>] {
+        return parentChildMap
```
I think I might need you to return me a a dictionary where the value is an array of child tags; from the UI side there might need to be a consistent ordering of the child tags under a parent tag

fterh replied:
This is by design. Within TagManager, I need to do membership check/add/remove operations efficiently, but TagManager does not maintain tags in any order. So an array doesn't make sense.

I think it'll be better if you convert to an array then sort haha

kumuwu replied:
As in, you can and should still internally maintain as a set, which will be unknown to me. I'm not sure if I should do any operations (i.e. sorting) on the tags? (separation of concern blah blah)
it's just like transactions internally are not stored in chronological order, but when UI gets hold of it it gets a sorted copy.
but, I'm not sure haha

fterh replied:
Okay, done in commit 67f01e5 :)

https://github.com/cs3217-crispybacon/bacon/pull/60/commits/67f01e532c40463a15b253a6ea0d00eb30e2bbfc

kumuwu commented on bacon/model/Tags/TagManager.swift:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    func getChildrenTagsOf(_ parent: String) throws -> Set<Tag> {
+        let parentTag = Tag(parent)
```
same here, an (sorted) array might be better

fterh committed:
Add TagManager tests for semi-singleton pattern

fterh committed:
Revise TagManager specifications

When adding a child Tag to a non-existent parent Tag, throw an error instead of automatically creating the parent Tag.

fterh committed:
Update specification to return sorted arrays instead of sets

fterh committed:
Conform Tag to Comparable

fterh committed:
Update TagManager implementation

Conform to new specifications

fterh committed:
Update TagManager tests
